cis194/week7/daniel-deng/src/CIS194/Homework07/JoinList.hs

```diff
@@ -0,0 +1,26 @@
+module CIS194.Homework07.JoinList ( JoinList(..) ) where
+
+import CIS194.Homework07.Sized ( Sized(..) )
+
+data JoinList m a = Empty
+                  | Single m a
+                  | Append m (JoinList m a) (JoinList m a)
+  deriving (Eq, Show)
+
+tag :: Monoid m => JoinList m a -> m
+tag Empty          = mempty
+tag (Single m _)   = m
+tag (Append m _ _) = m
+
+instance Monoid m => Semigroup (JoinList m a) where
```

Comment:
@shineli1984 I'm a bit confused here. 

```haskell
  Empty <> jl = jl
  jl <> Empty = jl
```

Shouldn't the definition of `mempty` take care of it when defining the `Monoid`?

I mean I understand that :

- `Semigroup` can be existing without `mempty`
- we need exhaustive pattern matching when defining `Semigroup` 
- we need that definition to declare the `Monad` instance

But does that mean basically any `Monoid` instances that needs pattern matching for `<>` needs to do left and right identity associativity? 

To me it's like saying:
```haskell
instance Semigroup A where
  (<>) :: A -> A -> A
  Empty <> x = x  -- Appending Empty with x would result in x itself
  x <> Empty = x  -- Appending x with Empty would result in x itself
  x1 <> x2 = blah -- Appending y1 and y2 would be whatever it would be

instance Monoid A where
  mempty :: A
  mempty = Empty -- BTW, deal readers, `Empty` is the identity element...
  
  -- So can haskell please take care of the left and right associativity for me?
```

That looks a bit verbose to me. Am I missing any language features? Or is that something that is just expected to be that verbose?

Reply:
Like you said, by `take care of the left and right ... for me`, GHC would have to produce non-exhaustive pattern matching. How do we make it error-proven and reduce the boilplate at same time is a question for the GHC folks. I don't want to take any guess here.